From 064c233d1e8bc05b338d0e237394936f3086ce13 Mon Sep 17 00:00:00 2001
From: Jouni Malinen <j@w1.fi>
Date: Thu, 4 Jan 2024 11:04:20 +0200
Subject: [PATCH] AP: Fix a regression in indoor 6 GHz AP determination

The change to use a helper function for checking he_6ghz_reg_pwr_type
missed the difference between two types of checks for different values:
indoor AP vs. SP AP. Fix this by introducing another helper function to
cover the indoor (i.e., SP and non-SP indoor cases).

Fixes: 121ccadeb44c ("AP: A helper function for determining whether the AP is an SP AP")
Signed-off-by: Jouni Malinen <j@w1.fi>
---
 src/ap/beacon.c              | 2 +-
 src/ap/ieee802_11.c          | 2 +-
 src/common/ieee802_11_defs.h | 6 ++++++
 3 files changed, 8 insertions(+), 2 deletions(-)

diff --git a/src/ap/beacon.c b/src/ap/beacon.c
index ee1f02792..89e360874 100644
--- a/src/ap/beacon.c
+++ b/src/ap/beacon.c
@@ -580,7 +580,7 @@ static size_t he_elem_len(struct hostapd_data *hapd)
 			3 + sizeof(struct ieee80211_he_6ghz_band_cap);
 		/* An additional Transmit Power Envelope element for
 		 * subordinate client */
-		if (he_reg_is_sp(hapd->iconf->he_6ghz_reg_pwr_type))
+		if (he_reg_is_indoor(hapd->iconf->he_6ghz_reg_pwr_type))
 			len += 4;
 
 		/* An additional Transmit Power Envelope element for
diff --git a/src/ap/ieee802_11.c b/src/ap/ieee802_11.c
index a939ad35e..50190a280 100644
--- a/src/ap/ieee802_11.c
+++ b/src/ap/ieee802_11.c
@@ -7077,7 +7077,7 @@ u8 * hostapd_eid_txpower_envelope(struct hostapd_data *hapd, u8 *eid)
 
 		/* Indoor Access Point must include an additional TPE for
 		 * subordinate devices */
-		if (he_reg_is_sp(iconf->he_6ghz_reg_pwr_type)) {
+		if (he_reg_is_indoor(iconf->he_6ghz_reg_pwr_type)) {
 			/* TODO: Extract PSD limits from channel data */
 			if (hapd->iconf->reg_sub_cli_eirp_psd != -1)
 				tx_pwr = hapd->iconf->reg_sub_cli_eirp_psd;
diff --git a/src/common/ieee802_11_defs.h b/src/common/ieee802_11_defs.h
index 4e040c35c..12137dcf4 100644
--- a/src/common/ieee802_11_defs.h
+++ b/src/common/ieee802_11_defs.h
@@ -2461,6 +2461,12 @@ enum he_reg_info_6ghz_ap_type {
 	HE_REG_INFO_6GHZ_AP_TYPE_MAX = HE_REG_INFO_6GHZ_AP_TYPE_INDOOR_SP,
 };
 
+static inline bool he_reg_is_indoor(enum he_reg_info_6ghz_ap_type type)
+{
+	return type == HE_REG_INFO_6GHZ_AP_TYPE_INDOOR ||
+		type == HE_REG_INFO_6GHZ_AP_TYPE_INDOOR_SP;
+}
+
 static inline bool he_reg_is_sp(enum he_reg_info_6ghz_ap_type type)
 {
 	return type == HE_REG_INFO_6GHZ_AP_TYPE_SP ||
-- 
2.43.0

